src/main/scala/zio/keeper/Cluster.scala

```diff
@@ -1,31 +1,267 @@
 package zio.keeper
 
-import java.net.InetAddress
+import java.math.BigInteger
+import java.util.UUID
 
-import zio.{ IO, ZIO }
+import zio._
+import zio.console.putStrLn
+import zio.keeper.Cluster.InternalProtocol.{ Iam, IamNewHere, PleaseIntroduceYourself }
+import zio.keeper.Error.{ HandshakeError, SendError }
+import zio.keeper.Error.NodeUnknown
+import zio.nio._
+import zio.nio.channels.{ AsynchronousServerSocketChannel, AsynchronousSocketChannel }
 import zio.stream.Stream
-import zio.keeper.Cluster._
 
-trait Cluster {
-  final def join: ZIO[Credentials with Discovery with Transport, Error, Unit] = ???
+import scala.reflect.ClassTag
+
+trait Cluster[A] {
```

Comment:
I might suggest to remove `A` and just support messages of type `Chunk[Byte]`. It simplifies the design and de-conflates serialization/deserialization from the protocol concerns.

Reply:
I agree let move this responsibility to library user.